### ISSUE TITLE
False-positive: Starting from version 4.0.0 false-positive on pytest-bdd fixtures

### ISSUE THREAD

**Issue by measles** [PyCQA/pylint#10665]
### Bug description

We have a set of `pytest_bdd` based tests, where every step is decorated with its description and has the same name `_`. Up until version 4.0.0, it was not a problem. But now every single step description after the first one emits a failed test `function-redefined`. Nothing was changed in a code and if I

### Configuration

```ini

```

### Command used

```shell
± uv run pylint -j 0 example.py
```

### Pylint output

```python
************* Module example
example.py:15:0: E0102: function already defined line 10 (function-redefined)
example.py:20:0: E0102: function already defined line 10 (function-redefined)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

### Expected behavior

Same as it was in versions before 4.0.0:
```
************* Module example
example.py:15:0: E0102: function already defined line 10 (function-redefined)
example.py:20:0: E0102: function already defined line 10 (function-redefined)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 0.00/10, +0.00)
```

This is an example for pytest version 3.3.9 on the very same file. File itself:

[example.py](https://github.com/user-attachments/files/22904683/example.py)

.

### Pylint version

```shell
debian:
± uv run  pylint --version
pylint 3.3.9
astroid 3.3.11
Python 3.13.5 (main, Jun 25 2025, 18:55:22) [GCC 14.2.0]

Same on Ubuntu:
± pylint --version
pylint 4.0.0
astroid 4.0.1
Python 3.13.5 (main, Jun 12 2025, 08:56:14) [GCC 13.3.0]
```

### OS / Environment

Debian:
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 13 (trixie)
Release:        13
Codename:       trixie

Ubuntu:
± lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 24.04.2 LTS
Release:        24.04
Codename:       noble

### Additional dependencies

```python

```

**Comment by DanielNoord** [comment]
This is because of the changes in https://github.com/pylint-dev/pylint/pull/10606

We considered this an improvement for most users. Could you provide the exact code that triggers these warnings?

**Comment by measles** [comment]
@DanielNoord [This is an example](https://github.com/user-attachments/files/22904683/example.py) of the file that triggers the code quality messages. Unfortunately, due to an NDA, I can't provide real code, but all the main details are in the example.

As these pytest-bdd fixtures will never be addressed by the function name, only by fixture description (like `@when(…)`, `@given(…)`, or `@then(…)`), so the entire repo has named only some utility functions or classes, as well as the scenario name (like in my example), all steps have `_` as a function name. 

Before the changes you've mentioned, it wasn't an issue. But for now I have only two ways: invest a lot of time in renaming to a unique name something that will never be addressed by this name or switch this specific check to all files that contain at least two step definitions. I've chosen the second option, but this means that currently all these files have no control over repetitive function names for the real functions that happen to occur in them. Not a full-scale disaster, but it makes a problem that I didn't have before `pytest` update.

**Comment by DanielNoord** [comment]
@jacobtylerwalls @Pierre-Sassoulas what do you think about this?

**Comment by jacobtylerwalls** [comment]
#10606 reverted #1369. I'd be open to reverting.

**Comment by Pierre-Sassoulas** [comment]
Nice catch, Jacob, I failed at reviewing this properly (line 83's removal was subtle 😬). Agree with the revert, but I wonder if it should be a partial revert with simply:
```
            dummy_variables_rgx = self.linter.config.dummy_variables_rgx
            if dummy_variables_rgx and dummy_variables_rgx.match(node.name):
                return
```
being added back in pylint/checkers/base/basic_error_checker.py.
